File: Bit manipulation/268. Missing Number.java
class Solution {
    /*
    XOR
    0 1 2 3
    0 1 3 4
    4^(0^0)^(1^1)^(2^3)^(3^4)
    = 2 
    SINCE (A^A) = A

    */
    public int missingNumber(int[] nums) {
        int res = nums.length;
        for(int i = 0; i < nums.length; i ++){
            res ^= nums[i]^i;
        }
        return res;
    }
}

File: Bit manipulation/268. Missing Number.cpp
class Solution
{
public:
    /*
    XOR each number in nums with the array that was supposed to be full from 0 to n

    */
    int missingNumber(vector<int> &nums)
    {
        int nums_size = nums.size();
        int missing = 0;
        for (int i = 0; i <= nums_size; i++)
        {
            missing = missing ^ i;
        }
        for (int i : nums)
        {
            missing = missing ^ i;
        }
        return missing;
    }
};

class Solution
{
public:
    int missingNumber(vector<int> &nums)
    {
        std::unordered_set<int> set;
        for (int n : nums)
        {
            set.insert(n);
        }
        for (int i = 0; i <= nums.size(); i++)
        {
            if (set.contains(i) == false)
            {
                return i;
            }
        }
        return -1;
    }
};